### DISCUSSION TITLE
ngtcp2 v1.4.0

### DISCUSSION THREAD

**Discussion by tatsuhiro-t** [ngtcp2/ngtcp2#1171] [Announcements]
## What's Changed
* Erase sensitive data before freeing memory by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1122
* Fix compile error with libstdc++6-14 by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1123
* Make congestion controller use the current path MTU by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1124
* Reduce malloc call in conn_new by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1125
* Add missing FindJemalloc.cmake to EXTRA_DIST by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1127
* Automate release process by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1130
* Make Path MTU Discovery probes configurable by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1128
* examples: Add --pmtud-probes option by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1132
* Accept zero length UDP datagram payload and just return 0 by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1134
* Deal with the case that send_quantum < max_udp_payload_size by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1135
* Adjust simpleclient buffer to have default max_tx_udp_payload_size by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1136
* Document about outgoing UDP datagram payload size by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1137
* Move ngtcp2_settings_default_versioned to ngtcp2_settings.c by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1138
* Refactor acktr by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1139
* Cleanup free functions called from conn_new by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1140
* Make functions that discard pkns callable from the other source files by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1141
* Add typed ngtcp2_min and ngtcp2_max functions by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1142
* Avoid setting 0 after memset by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1143
* Move ngtcp2_transport_params functions to its own file by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1144
* Remove unused ngtcp2_conversion_test.c by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1145
* Move struct version to the last argument by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1146
* git clone recursive by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1147
* Update README.rst by @Karthikdasari0423 in https://github.com/ngtcp2/ngtcp2/pull/1150
* ngtcp2_conn_write_connection_close: Fix assertion failure by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1154
* Fix assertion failure because of failing dup Connection ID check by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1155
* fuzz: Add read_write_pkt fuzzer by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1156
* Workaround llvm issue by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1158
* fuzz: Add missing include by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1159
* fuzz: Workaround llvm issue by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1160
* Add 2 new ngtcp2_ccerr_type values by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1161
* Add handshake fuzzer by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1162
* docker: Use copy --link by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1163
* Bump aws-lc to v1.23.0 by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1164
* Bump boringssl by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1165
* Bump picotls by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1166
* Switch to distroless/base-nossl by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1167
* Remove debug printf by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1168
* Add padding to at most 1200 bytes by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1169
* Add ngtcp2_ppe padding tests by @tatsuhiro-t in https://github.com/ngtcp2/ngtcp2/pull/1170

## New Contributors
* @Karthikdasari0423 made their first contribution in https://github.com/ngtcp2/ngtcp2/pull/1150

**Full Changelog**: https://github.com/ngtcp2/ngtcp2/compare/v1.3.0...v1.4.0

> [!IMPORTANT]
>
> The asc files are now signed with `rsa4096/5339A2BE82E07DEC`.  You can find it at keyserver.ubuntu.com.

<hr /><em>This discussion was created from the release <a href='https://github.com/ngtcp2/ngtcp2/releases/tag/v1.4.0'>ngtcp2 v1.4.0</a>.</em>

**Comment by devkdas** [comment]
Thanks @tatsuhiro-t

**Comment by teddysun** [comment]
The below file looks like broke. file size is too small.
[ngtcp2-1.4.0.tar.gz](https://github.com/ngtcp2/ngtcp2/releases/download/v1.4.0/ngtcp2-1.4.0.tar.gz)
[ngtcp2-1.4.0.tar.xz](https://github.com/ngtcp2/ngtcp2/releases/download/v1.4.0/ngtcp2-1.4.0.tar.xz)

**Comment by tatsuhiro-t** [comment]
I'll fix tar balls.  For now those broken files are removed.

**Comment by tatsuhiro-t** [comment]
Fixed now.

**Comment by vszakats** [comment]
Thanks for the digital signature, it's very welcome.

But, the new signing key has caused some problems in curl-for-win. ~~The fix that worked for me was to also import the old parent key `F4F3B91474D1EB29889BD0EF7E8403D5D673C366` (DSA1024). Without it, gpg is saying: `Can't check signature: No public key`~~

~~Maybe because the new key ([`516b622918d15c478ab1ea3a5339a2be82e07dec`](https://keyserver.ubuntu.com/pks/lookup?search=5339A2BE82E07DEC&fingerprint=on&op=index)) is a subkey?~~

The other problem is that pulling these keys from a non-Ubuntu server didn't seem to work. https://pgpkeys.eu/ doesn't have them, https://keys.openpgp.org/ also didn't work for some reason. Speaking of a new key, maybe they need more time to replicate there?

~~I'm wondering if it would help if the new key would be a primary key, not a subkey?~~

**Reply by vszakats** [reply]
update: may bad, it now works with the subkey only, but requires a direct key download from Ubuntu: https://ci.appveyor.com/project/curlorg/curl-for-win/builds/49427252#L2591

**Reply by tatsuhiro-t** [reply]
It is indeed subkey.  I automated release process on github actions and it is too risky to upload parent key.

**Reply by vszakats** [reply]
Being a subkey is OK, it looks like! The failure was due to `keys.openpgp.org` [returning](https://keys.openpgp.org/pks/lookup?op=get&options=mr&exact=on&search=0x516B622918D15C478AB1EA3A5339A2BE82E07DEC) the subkey, but not in a usable form.

**Reply by vszakats** [reply]
Compared to the Ubuntu server, `keys.openpgp.org` is returning a key that's missing a 'User ID' packet (and a signature packet with it).

**Reply by tatsuhiro-t** [reply]
I heard that openpgp.org server does not return user id unless it is explicitly allowed by a user.  I do no know how to do that.

**Reply by vszakats** [reply]
Might be privacy reasons. The behaviour is indeed not specific to this key. I'll drop this server from the list (added in January without much testing). It will fix the problem by falling back to Ubuntu's automatically.

Hopefully the key server [`pgpkeys.eu`](https://pgpkeys.eu/pks/lookup?search=F4F3B91474D1EB29889BD0EF7E8403D5D673C366&fingerprint=on&op=index) will eventually get refreshed with the new subkey, so not even the fallback will be necessary.

**Reply by tatsuhiro-t** [reply]
I can send my key to pgpkeys.su if that helps.

**Reply by vszakats** [reply]
Thanks, I think it would be useful, probably for others too. Notice the TLD is `.eu` (not `.su`).

**Reply by tatsuhiro-t** [reply]
I sent new key to pgpkeys.eu

**Reply by vszakats** [reply]
Thank you!
